csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -0,0 +1,149 @@
+
+         /// <summary>When true, the operation is idempotent.</summary>
+        public bool IsIdempotent { get; }
+
+        /// <summary>The request context. Its initial value is computed when the request frame is created.</summary>
+        public Context Context { get; }
+
+        /// <summary>An OutputStream is sealed when it can no longer be written into, in particular the payload
+        /// is complete and the context is written. Modidying Context once IsSealed is true has no effect.</summary>
+        public bool IsSealed { get; private set; } = false; // TODO: move to OutputStream with protected set
+
+        /// <summary>Creates a new outgoing request frame with an empty payload for the in-parameters.</summary>
+        /// <param name="proxy">A proxy to the target Ice object. This method uses the communicator, identity, facet,
+        /// encoding and context of this proxy to create the request frame.</param>
+        /// <param name="operation">The operation to invoke on the target Ice object.</param>
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame with an empty payload.</returns>
+        public static OutgoingRequestFrame Empty(IObjectPrx proxy, string operation, bool idempotent,
+            Context? context = null) => Create(proxy, operation, idempotent, ArraySegment<byte>.Empty, context);
+
+        /// <summary>Creates a new outgoing request frame and starts the payload (encapsulation) that will hold the
+        /// in-parameters in that frame.</summary>
+        /// <param name="proxy">A proxy to the target Ice object. This method uses the communicator, identity, facet,
+        /// encoding and context of this proxy to create the request frame.</param>
+        /// <param name="operation">The operation to invoke on the target Ice object.</param>
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="format">The Slice format (Compact or Sliced) used by the encapsulation.</param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame.</returns>
+        public static OutgoingRequestFrame Start(IObjectPrx proxy, string operation, bool idempotent,
+            FormatType? format = null, Context? context = null)
+        {
+            var frame = new OutgoingRequestFrame(proxy.Communicator, proxy.Identity, proxy.Facet, operation, idempotent,
+                proxy.Context, context);
+            frame.StartEncapsulation(proxy.EncodingVersion, format);
```

Comment:
What is the idea here, when we integrate this with vectored IO, the payload starts in a new segment?

Reply:
I think would be nice to have matching names for Start/End method names, like other Start/End methods

`CreateAndStartPayload` is probably too long

Reply:
Fixed as per Joe's suggestion to use constructors and a delegate to write the payload.